public/app/features/explore/PrometheusListView/RawListContainer.tsx

```diff
@@ -0,0 +1,162 @@
+
+export interface RawListContainerProps {
+  tableResult: DataFrame;
+}
+
+const getRawListContainerStyles = stylesFactory(() => {
+  return {
+    wrapper: css`
+      height: 100%;
+      overflow: scroll;
+    `,
+    mobileWrapper: css`
+      height: 100%;
+      overflow: scroll;
+    `,
+    header: css`
+      display: flex;
+      justify-content: space-between;
+      align-items: center;
+      padding: 10px 0;
+    `,
+  };
+});
+
+const mobileWidthThreshold = 480;
+const numberOfColumnsBeforeExpandedViewIsDefault = 2;
+
+/**
+ * The container that provides the virtualized list to the child components
+ * @param props
+ * @constructor
+ */
+const RawListContainer = (props: RawListContainerProps) => {
+  const { tableResult } = props;
+  let dataFrame = cloneDeep(tableResult);
+  const styles = getRawListContainerStyles();
+  let listRef = useRef<List | null>(null);
+
+  const valueLabels = dataFrame.fields.filter((field) => field.name.includes('Value'));
+  const items = getRawPrometheusListItemsFromDataFrame(dataFrame);
+  const { width } = useWindowSize();
+  const [isExpandedView, setIsExpandedView] = useState(
+    width <= mobileWidthThreshold || valueLabels.length > numberOfColumnsBeforeExpandedViewIsDefault
+  );
+
+  const onContentClick = () => {
+    setIsExpandedView(!isExpandedView);
+  };
+
+  useEffect(() => {
+    // After the expanded view has updated, tell the list to re-render
+    listRef.current?.resetAfterIndex(0, true);
+  }, [isExpandedView]);
+
+  const calculateInitialHeight = (length: number): number => {
+    const maxListHeight = 600;
+    const shortListLength = 10;
+
+    if (length < shortListLength) {
+      let sum = 0;
+      for (let i = 0; i < length; i++) {
+        sum += getListItemHeight(i, true);
+      }
+
+      return Math.min(maxListHeight, sum);
+    }
+
+    return maxListHeight;
+  };
+
+  const getListItemHeight = (itemIndex: number, isExpandedView: boolean) => {
+    const singleLineHeight = 32;
+    const additionalLineHeight = 22;
+    if (!isExpandedView) {
+      return singleLineHeight;
+    }
+    const item = items[itemIndex];
+
+    // Height of 1.5 lines, plus the number of non-value attributes times the height of additional lines
+    return 1.5 * singleLineHeight + (Object.keys(item).length - valueLabels.length) * additionalLineHeight;
+  };
+
+  return (
+    <section>
+      <header className={styles.header}>
+        <Button
+          aria-hidden="true"
+          variant="secondary"
+          role={'button'}
+          title={`${isExpandedView ? 'Contract results' : 'Expand results'}`}
+          icon={isExpandedView ? 'minus' : 'plus'}
+          type="button"
+          onClick={onContentClick}
+        >
+          {isExpandedView ? 'Contract' : 'Expand'}
+        </Button>
+        <div>Result series: {items.length}</div>
+      </header>
+
+      <div role={'table'}>
```

Comment:
i see you are using divs with roles table/row/cell. what is the advantage over using plain `table`, `tr`s and `td`s? can't we just use them instead?

Reply:
It's not really a valid table with all the extra markup from the virtual list component, I tried implementing this with our virtual table component, but couldn't get it to provide the functionality we were looking for here, so I stuck with modifying the display of the virtual list component, the data is just tabular enough that I added the extra roles (but mostly to facilitate unit tests).